### ISSUE TITLE
[BUG]When loading from checkpoint to continue training, it will hang during the validation‘s forward.

### ISSUE THREAD

**Issue by young-chao** [NVIDIA/Megatron-LM#647]
**Describe the bug**
When I load the model from checkpoint and continue training, it always hangs during the validation process. It should be noted that it only hangs when the "--no-load-optim" and "--no-load-rng" parameters are not used. 

**To Reproduce**
Below is my startup script：
`
torchrun --nnodes=2 --nproc_per_node=8 --node_rank=$node_rank \
    --master_addr=10.90.0.2 --master_port=12375  \
    anypath/Megatron-LM/pretrain_gpt.py \
    --use-checkpoint-args \
    --exit-on-missing-checkpoint \
    --load=anypath/megatron/13B_10B \
    --save=anypath/megatron/13B_10B \
    --train-data-path=anypath/CC_megatron_merged \
    --valid-data-path=anypath/CC2_text_document \
    --tensor-model-parallel-size=2 \
    --pipeline-model-parallel-size=4 \
    --recompute-activations \
    --use-distributed-optimizer \
    --sequence-parallel \
    --use-flash-attn \
    --overlap-grad-reduce \
    --bf16 \
    --num-layers=40 \
    --hidden-size=5120 \
    --num-attention-heads=40 \
    --ffn-hidden-size=13824 \
    --attention-dropout=0.0 \
    --hidden-dropout=0.0 \
    --swiglu \
    --normalization RMSNorm \
    --use-rotary-position-embeddings \
    --no-position-embedding \
    --no-masked-softmax-fusion \
    --untie-embeddings-and-output-weights \
    --disable-bias-linear \
    --tokenizer-type=Llama2Tokenizer \
    --tokenizer-model=anypath/tokenizer.model \
    --train-iters=160000 \
    --micro-batch-size=1 \
    --global-batch-size=32 \
    --seq-length=2048 \
    --lr=3e-4 \
    --min-lr=0 \
    --lr-decay-style=linear \
    --adam-beta1=0.9 \
    --adam-beta2=0.95 \
    --weight-decay=0.1 \
    --clip-grad=1.0 \
    --init-method-std=0.006 \
    --lr-decay-iters=157709 \
    --lr-warmup-iters=15771 \
    --max-position-embeddings=4096 \
    --log-interval=1 \
    --eval-interval=5 \
    --eval-iters=1 \
    --save-interval=1000 \
    --tensorboard-queue-size=1 \
    --tensorboard-dir=anypath \
    --log-timers-to-tensorboard \
    --log-batch-size-to-tensorboard \
    --log-validation-ppl-to-tensorboard \
    --num-workers=8 \
    --seed=1234
`

**Expected behavior**
When the program loads from checkpoint and continues training for 5 steps, it will hang until it times out.

**Environment (please complete the following information):**
I used the image from [NGC](https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/rel-23-07.html), and upgraded nccl version.
 - Megatron-LM [2bc6cd3](https://github.com/NVIDIA/Megatron-LM/commit/2bc6cd307a11423928c675f741e79e03df23e721)
 - PyTorch 2.1.0a0+b5021ba
 - CUDA 12.1.1
 - NCCL 2.18.6

**Additional context**
It is worth noting that this error is sporadic. When training on two machines, training can continue normally in about half of the cases. However, when training on 64 machines, it will hang after trying many times.

**Comment by young-chao** [comment]
What I found was that the training got stuck in the first or last stage of PP where it was trying to get the data.
![image](https://github.com/NVIDIA/Megatron-LM/assets/34190033/3cffacaf-ce33-41b5-a230-f3a7d563df8a)

**Comment by young-chao** [comment]
你好，你的邮件已收到，谢谢！

**Comment by young-chao** [comment]
你好，你的邮件已收到，谢谢！

**Comment by young-chao** [comment]
你好，你的邮件已收到，谢谢！

**Comment by young-chao** [comment]
你好，你的邮件已收到，谢谢！
